### PR TITLE
Resolved #1639: display sni on ClientConnection

### LINKED ISSUE #1639: Display Server Name Indication in the UI
##### Steps to reproduce the problem:

1. `mitmproxy`
2. `curl -x localhost:8080 https://example.com/`
3. View Details Tab

##### What is the expected behavior?

The Server Name Indication sent by the client should be displayed in the UI.

##### What went wrong?

The SNI is not displayed in the "Client Connection" section.

##### Any other comments? What have you tried so far?

The first step here is to make a `.sni` attribute in ClientConnection which also gets serialized. I'm happy to help if anyone wants to pick this issue up! :smiley:

---

Mitmproxy Version: master
Operating System: Win10 x64

### PR DESCRIPTION
**Pull request by chhsiao90**
There is still one test cannot pass that I need your advise how could I solved that.

The test is ```test/mitmproxy/test_flow_format_compat.py``` that will load a dump file ```data/dumpfile-011```
The dumpfile didn't contains sni info so that when the program try deserialize from dumpfile,
the state of ClientConnection didn't contains the sni field.

Is that I need add sni in the dumpfile manually? Or there is another way?
Thanks!

```
    def test_load():
        with open(tutils.test_data.path("data/dumpfile-011"), "rb") as f:
            flow_reader = io.FlowReader(f)
>           flows = list(flow_reader.stream())

test/mitmproxy/test_flow_format_compat.py:9: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
mitmproxy/io.py:43: in stream
    yield FLOW_TYPES[data["type"]].from_state(data)
mitmproxy/flow.py:112: in from_state
    f.set_state(state)
mitmproxy/flow.py:107: in set_state
    super().set_state(state)
mitmproxy/stateobject.py:60: in set_state
    obj = cls.from_state(val)
mitmproxy/connections.py:81: in from_state
    f.set_state(state)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ClientConnection: [ssl] None>, state = {'address': {'address': [b'127.0.0.1', 60786], 'use_ipv6': False}, 'timestamp_end': None, 'timestamp_start': 1469208633.936}

    def set_state(self, state):
        """
            Load object state from data returned by a get_state call.
            """
        state = state.copy()
        for attr, cls in self._stateobject_attributes.items():
>           val = state.pop(attr)
E           KeyError: 'sni'

mitmproxy/stateobject.py:52: KeyError
```

### CODE REVIEW

**Comment by chhsiao90** [comment]
refs #1639

File: mitmproxy/proxy/protocol/tls.py [hunk 1]
```diff
@@ -336,6 +336,7 @@ def __call__(self):
             # Peek into the connection, read the initial client hello and parse it to obtain SNI and ALPN values.
             try:
                 self._client_hello = TlsClientHello.from_client_conn(self.client_conn)
+                self.client_conn.sni = self._client_hello.sni
```
Comment (mhils) [inline mitmproxy/proxy/protocol/tls.py]: This is using our own TlsClientHello parser, which may not work under some circumstances. Can we use [`.get_servername()`](http://www.pyopenssl.org/en/stable/api/ssl.html?highlight=get_servername#OpenSSL.SSL.Connection.get_servername) [here](https://github.com/chhsiao90/mitmproxy/blob/7274b3a58f8eb36e68bc4d758c5e191b684d16d9/mitmproxy/proxy/protocol/tls.py#L485) instead?

**Comment by mhils** [comment]
(Note to whoever merges this: please add https://github.com/mitmproxy/mitmproxy/blob/21f133fae9faf16f5b4322fab3e92eaa8a2bb719/mitmproxy/io_compat.py#L58 to the 0.19 converter afterwards. I refrain from doing that right away to avoid unecessary merge conflicts)

File: mitmproxy/proxy/protocol/tls.py (hunk 1 above)
Comment (chhsiao90) [inline mitmproxy/proxy/protocol/tls.py]: Got it, I will update it.

**Comment by chhsiao90** [comment]
Had fixed about how we get the sni, and io_compat.py for versions compatibility of dumpfile.
Please help review it, thanks!
I will help on #582 when I'm free and if there anyone else pick it up~

**Comment by mhils** [comment]
That was a quick turnaround - look great, thank you very much! 😊

**Pull request merged**